proxyd/consensus_aware.go

```diff
@@ -0,0 +1,392 @@
+				continue
+			}
+
+			actualBlockNumber, actualBlockHash, err := cp.fetchBlock(ctx, be, proposedBlock)
+			if err != nil {
+				log.Warn(fmt.Sprintf("error updating backend [%s]: %v", be.Name, err))
+				continue
+			}
+			if proposedBlockHash == "" {
+				proposedBlockHash = actualBlockHash
+			}
+			blocksDontMatch := (actualBlockNumber != proposedBlock) || (actualBlockHash != proposedBlockHash)
+			if blocksDontMatch {
+				if currentConsensusBlockNumber >= actualBlockNumber {
+					log.Warn(fmt.Sprintf("backend [%s] broke consensus, block number: %s (expected %s), block hash: %s (expected %s)", be.Name, actualBlockNumber, proposedBlock, actualBlockHash, proposedBlockHash))
+					broken = true
+				}
+				allAgreed = false
+				break
+			}
+			consensusBackends = append(consensusBackends, be)
+			consensusBackendsNames = append(consensusBackendsNames, be.Name)
+		}
+		if allAgreed {
+			hasConsensus = true
+		} else {
+			// walk one block behind and try again
+			proposedBlock = hexAdd(proposedBlock, -1)
+			proposedBlockHash = ""
+			log.Info(fmt.Sprintf("no consensus, now trying block: %s", proposedBlock))
+		}
+	}
+
+	if broken {
+		// propagate event to other interested parts, such as cache invalidator
+		log.Info(fmt.Sprintf("consensus broken, was at: %s, now at: %s [%s]", currentConsensusBlockNumber, proposedBlock, proposedBlockHash))
+	}
+
+	cp.Tracker.SetConsensusBlockNumber(proposedBlock)
+	consensusLatestBlock.Set(blockToFloat(proposedBlock))
+	cp.consensusGroupMux.Lock()
+	cp.ConsensusGroup = consensusBackends
+	cp.consensusGroupMux.Unlock()
+
+	log.Info(fmt.Sprintf("group state: %s, consensusBackends: [%s], filteredBackendsNames: [%s]", proposedBlock, strings.Join(consensusBackendsNames, ", "), strings.Join(filteredBackendsNames, ", ")))
+}
+
+func (cp *ConsensusPoller) fetchBlock(ctx context.Context, be *Backend, block string) (blockNumber string, blockHash string, err error) {
+	rpc := fmt.Sprintf("{\"jsonrpc\":\"2.0\",\"method\":\"eth_getBlockByNumber\",\"params\":[\"%s\", false],\"id\":67}", block)
+	var rpcReq RPCReq
+	err = json.Unmarshal([]byte(rpc), &rpcReq)
+	if err != nil {
+		return "", "", fmt.Errorf(fmt.Sprintf("error unmarshalling rpcreq on backend %s", be.Name), err)
+	}
+	res, err := be.doForward(ctx, []*RPCReq{&rpcReq}, false)
+	if err != nil {
+		return "", "", fmt.Errorf(fmt.Sprintf("error forwarding rpc call to check consensus on backend %s", be.Name), err)
+	}
+	if len(res) != 1 {
+		return "", "", fmt.Errorf(fmt.Sprintf("unexpected response len checking consensus on backend (len != 1) for backend %s", be.Name), err)
+	}
+	if res[0].IsError() {
+		return "", "", fmt.Errorf(fmt.Sprintf("error checking consensus on backend %s", be.Name), res[0].Error)
+	}
+	jsonMap, ok := res[0].Result.(map[string]interface{})
+	if !ok {
+		return "", "", fmt.Errorf(fmt.Sprintf("unexpected response type checking consensus on backend %s", be.Name))
+	}
+	blockNumber = jsonMap["number"].(string)
```

Comment:
Since the node `eth_getBlockByNumber` outputs the block number in hex, and only take the input as an hex encoded string, I rather use this internal representation.

We only convert to decimal to decrement the block number.

Reply:
Check out `hexutil.Uint64`